Sources/AsyncAlgorithms/AsyncMapErrorSequence.swift

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public extension AsyncSequence {
+
+    /// Converts any failure into a new error.
+    ///
+    /// - Parameter transform: A closure that takes the failure as a parameter and returns a new error.
+    /// - Returns: An asynchronous sequence that maps the error thrown into the one produced by the transform closure.
+    ///
+    /// Use the ``mapError(_:)`` operator when you need to replace one error type with another.
+    func mapError<ErrorType>(transform: @Sendable @escaping (Error) -> ErrorType) -> AsyncMapErrorSequence<Self, ErrorType> {
```

Comment:
Should we return a `some AsyncSequence` instead of a concrete one here as well?

Reply:
Yeah they predate `AsyncSequence` adopting primary associated types so we can do that now. It should also allow us to avoid making the actual types public

Reply:
I decided not to return `some AsyncSequence` and return concrete type instead because using an opaque type would render typed throws ineffective by erasing the type, thereby preventing the compiler from ensuring that the returned sequence matches our intended new type. Consequently, it would simply default to any Error. And I believe the benefits of using typed throws for this specific case outweight the exposure of the internal types.

Additionally, I didn't use `Failure` since it's only available in iOS 18+, I'd need 2 functions to make it backwards compatible. And `AsyncSequence.Failure` is actuall `any Error` - there's no benefit of using `Failure` v.s. `any Error` since they are the same. (side note: `AsyncIteratorProtocol.next()` didn't adopt typed throws but `AsyncIteratorProtocol.next(isolation:)` does adopt typed throws. Looks like intentional to maintain backward compatibility. But it's also one of the reasons why I didn't use `Failure`.) You can checkout this commit to see my attempt https://github.com/clive819/swift-async-algorithms/commit/a8f139a49dc9efa7657993dd52782d98ed9c6d4b#diff-39026a27cb7e7c71fcd5df08416cb255085e46b3a227abecdc820d13a0307ab0R12-R15

Reply:
Address in https://github.com/apple/swift-async-algorithms/pull/324/commits/4ba9c84becdc3279ced63bf2faea6ce705a9769d

Also created a post for it https://forums.swift.org/t/pitch-map-error/75422